Battle Royal.playground/Contents.swift

```diff
@@ -0,0 +1,488 @@
+import UIKit
+
+protocol Resource{
+    var name: String { get }
+    
+    var damageBonus: Int { get }
+    var healBonus: Int {get}
+}
+
+protocol Unit{
+    var hp: Int { get set }
+    var damage: Int { get }
+    var luck: Int { get }
+    var haveUniqueSkill: Bool {get set}
+    
+    var resource: Resource? {get set}
+    
+    func hit(player: Player)
+    func findResource()
+    
+    func voiceOfDeath()
+    func voiceOfKilling()
+    func voiceOfHavenotSkill()
+    func voiceOfFind()
+}
+
+class MagicKnife: Resource{
+    var name: String = "MagicKnife"
+    
+    var damageBonus: Int = 10
+    
+    var healBonus: Int = 0
+    
+}
+
+class Medikit: Resource{
+    var name: String = "Medikit"
+    
+    var damageBonus: Int = 0
+    
+    var healBonus: Int = 10
+    
+    
+}
+
+class Armor: Resource{
+    var name: String = "Armor"
+    
+    var damageBonus: Int = 4
+    
+    var healBonus: Int = 5
+    
+    
+}
+
+class Knight: Unit{
+    
+    var hp: Int = 50
+    
+    var damage: Int = 15
+    
+    var luck: Int = 3
+    
+    var haveUniqueSkill: Bool = true
+    
+    var resource: Resource? = nil
+    
+    
+    func hit(player: Player) {
+        if resource == nil{
+            player.person.hp = player.person.hp - Int.random(in: damage...damage+10)
+        }
+        else {
+            player.person.hp = player.person.hp - Int.random(in: damage...damage+10)+resource!.damageBonus
+        }
```

Comment:
Можно сильно сократить. Если используешь optional, можно использовать конструкцию "`value` ?? `defaultValue`"

```suggestion
        player.person.hp = player.person.hp - Int.random(in: damage...damage+10)+(resource?.damageBonus ?? 0)
```